test/test_jit_py3.py

```diff
@@ -533,6 +533,22 @@ def forward(self, x):
         # Check that ignored method is still intact.
         self.assertEqual(inp, n.ignored_method(inp))
 
+    def test_if_returning_any(self):
+        """
+        Check that an if statement can return different
+        types early from each branch when the return
+        type of the function is Any.
+        """
+        def if_function(inp: torch.Tensor) -> Any:
+            if inp.shape[0] == 1:
+                return inp * inp
+            else:
+                return 3
+
+
+        with torch._jit_internal._disable_emit_hooks():
```

Comment:
An unintended consequence of this change is that IR -> Python conversion broke because we always generate Python for an if node that looks like this:
```
if ...:
  ...
  a = 1
else:
  a = 2

return a
```

This doesn't work for this case because `a` cannot have two separate types in two separate code paths. I couldn't find a quick and easy way to modify Python code generation to account for this so I disabled export/import in the test for now while we discuss.

The only way I see around this is to hoist the return into the if and else branches if it comes immediately after the if. What do you think?

```
if ...:
  ...
  return 1
else:
  return 2
```

Reply:
I'm not sure what the best solution is here. Potentially we could explicitly annotate a: Any in some cases

Reply:
It should be possible to tell when a value is used as an Any output of an `If`, and to add an explicit annotation.

The simple case is that it only has one use, and to annotate the value with `Any` when it's printed. The trickier case is:
```
if cond:
    x = 1
    y = x + x
    return x, y
else:
    return "str", 3
```
Where the return value needs to be an int for its other uses, and then up-casted to an Any as the if-output. Maybe we could print this as 
if cond:
    x = 1
    y = x + x
    out = unchecked_cast<Any, x>, y
else:
    out = "str", 3
return out
```

Reply:
Is that code legal? Something like this is not:
```
def my_fn(...) -> Any:
  a: Any = 3
  if cond:
    a = 5
  else:
    a = "five"
  return a
```
but something like this is:
```
def my_func(...) -> Any:
  a: Any = 3
  if cond:
    return 5
  else:
    return "five"
```

Reply:
yea, you're right the first example is not legal

Reply:
how does that affect my comments ?

Reply:
I just wanted to make the point that annotating the output with `Any` is not sufficient, we need to emit the `return` inside the `if/else` branches, which we don't currently do.

Reply:
The problem in your example is that neither if output has Any type, which I don't htink applies to my suggestion, that's what the `unchecked_cast<Any>` is for

Reply:
thanks so much for helping us on this @SplitInfinity @eellison ! Just an idea: would it help resolve the issue here if we require user to use an explicit `cast_to_any` API if they want to unify return types to `Any`? Something like:
```
def my_func(...) -> Any:
  if cond:
    return cast_to_any(5)
  else:
    return cast_to_any("five")
```

Reply:
@yf225 that would be a user-work around but I think we should just fix the problem in the compiler